### PR TITLE
Change InstantiateType to be an Action

### LINKED ISSUE #2594: Explorer: change `InstantiateType` to be an `Action` instead of a method
### Background

Our current explorer implementation uses a simple function for type instantiation (substitute placeholder type with actual types). A better approach would be to have it be an `Action` for the action stack.

### Goal

`InstantiateType` replaces all type variables (for example `T`) by their actual type values (for example `i32`). The goal is to create a new `TypeInstantiationAction` (`class TypeInstantiationAction : public Action`) that produces the same behavior.

The `InstantiateType` method is declared in [interpreter.cpp:139](https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/interpreter.cpp#L139)
```cpp
  // Instantiate a type by replacing all type variables that occur inside the
  // type by the current values of those variables.
  //
  // For example, suppose T=i32 and U=bool. Then
  //     __Fn (Point(T)) -> Point(U)
  // becomes
  //     __Fn (Point(i32)) -> Point(bool)
  //
  // TODO: This should be an Action.
  auto InstantiateType(Nonnull<const Value*> type, SourceLocation source_loc)
      -> ErrorOr<Nonnull<const Value*>>;
```

### Suggested approach

At a high-level, this could be achieved by:

1. Adding a new `TypeInstantiationAction` enum to `Action::Kind`
1. Creating a new `TypeInstantiationAction` class, following the example of similar classes like `ExpressionAction`
1. Updating code that makes use of `Action`s:
    * `Interpreter::Step`: entry point of actions, to invoke `InstantiateType` method
    * `Action::Print`: prints an action
1. Changing code that directly invokes `InstantiateType`, to instead spawn a new `TypeInstantiationAction`
    * For example: `return todo_.Spawn(std::make_unique<TypeInstantiationAction>(...));`
    * This will require using `act.pos()` to add a new step for instantiation, before dependent code can be executed. See [Interpreter readme](https://github.com/carbon-language/carbon-lang/tree/trunk/explorer/interpreter#explorer-execution) for a description of the action stack

### Resources

* https://github.com/carbon-language/carbon-lang/blob/trunk/CONTRIBUTING.md
* https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md
* `#contributing-help` channel on the Discord

### PR DESCRIPTION
**Pull request by prabhatexit0**
Added an alternative way of Instantiating types (it was previously done using `InstantiateType` method).
Created an `Action` called `TypeInstantiationAction` and changed most of the code that directly invokes `InstantiateType` to instead spawn a new `TypeInstantiationAction`.

Relates to [#2594 ](https://github.com/carbon-language/carbon-lang/issues/2594)

### CODE REVIEW

File: explorer/interpreter/interpreter.cpp [hunk 1]
```diff
@@ -95,6 +95,8 @@ class Interpreter {
   auto StepCleanUp() -> ErrorOr<Success>;
   auto StepDestroy() -> ErrorOr<Success>;
 
+  auto StepInstantiateType() -> ErrorOr<Success>;
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp:98]:
```suggestion
  // State transitions for type instantiation.
  auto StepInstantiateType() -> ErrorOr<Success>;
```

File: explorer/interpreter/interpreter.cpp [hunk 2]
```diff
@@ -1150,10 +1221,8 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
       } else {
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp:1167]: You will have to modify the sequence here to fit the new action.
Pseudo code below go give an example

```cpp
      bool interface_member = access.impl().has_value() && !forming_member_name;
      ...
      if (act.pos() == 0) {...}
      // Would normally be 2, but because the condition above is `act.pos() == 1 && foo`, we will have to approach that differently
      } else if (act.pos() == 1 && interface_member || act.pos() == 2) {
        // ... Instantiate type
      } else {
        // Get result from instantiate type, likely with `act.results()[n]` here, not sure at which index it is stored here
        return todo_FinishAction(act.results()[0]);
      }
        
      }
```

File: explorer/interpreter/interpreter.cpp [hunk 3]
```diff
@@ -1150,10 +1221,8 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
         // Finally, produce the result.
         if (auto constant_value = access.constant_value()) {
-          CARBON_ASSIGN_OR_RETURN(
-              Nonnull<const Value*> instantiated,
-              InstantiateType(*constant_value, access.source_loc()));
-          return todo_.FinishAction(instantiated);
+          return todo_.Spawn(std::make_unique<TypeInstantiationAction>(
+              *constant_value, access.source_loc()));
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: That's the idea yes! See comment above

Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: *Read the comment above.*

I think I got what I need to do but I'm confused at few places, will understand more about it, figure out the sequence that will be required for different cases and will commit the changes here.

Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: How should I approach this, when `InstantiateType` is called from a **non-step** method?
Should I get the current action by this: `auto act = todo_.CurrentAction();`
and write that sequence (spawning the action, then getting the value of it using different value of`act.pos()`)?
but this method won't be called again like a Step does for state transition right?

```cpp
auto Interpreter::EvalAssociatedConstant(
    Nonnull<const AssociatedConstant*> assoc, SourceLocation source_loc)
    -> ErrorOr<Nonnull<const Value*>> {
  // Instantiate the associated constant.
  CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> interface,
                          InstantiateType(&assoc->interface(), source_loc));
  CARBON_ASSIGN_OR_RETURN(Nonnull<const Witness*> witness,
                          InstantiateWitness(&assoc->witness()));

  const auto* impl_witness = dyn_cast<ImplWitness>(witness);
  if (!impl_witness) {
    CARBON_CHECK(phase() == Phase::CompileTime)
        << "symbolic witnesses should only be formed at compile time";
    CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> base,
                            InstantiateType(&assoc->base(), source_loc));
    return arena_->New<AssociatedConstant>(base, cast<InterfaceType>(interface),
                                           &assoc->constant(), witness);
  }

  // We have an impl. Extract the value from it.
  Nonnull<const ConstraintType*> constraint =
      impl_witness->declaration().constraint_type();
  std::optional<Nonnull<const Value*>> result;
  for (const auto& rewrite : constraint->rewrite_constraints()) {
    if (&rewrite.constant->constant() == &assoc->constant() &&
        TypeEqual(&rewrite.constant->interface(), interface, std::nullopt)) {
      // TODO: The value might depend on the parameters of the impl. We need to
      // substitute impl_witness->type_args() into the value.
      result = rewrite.converted_replacement;
      break;
    }
  }
  if (!result) {
    CARBON_FATAL() << impl_witness->declaration() << " with constraint "
                   << *constraint
                   << " is missing value for associated constant "
                   << *interface << "." << assoc->constant().binding().name();
  }
  return *result;
}
```

File: explorer/interpreter/interpreter.cpp [hunk 4]
```diff
@@ -423,30 +424,32 @@ auto Interpreter::StepLvalue() -> ErrorOr<Success> {
...
         return todo_.Spawn(std::make_unique<LValAction>(&access.object()));
-      } else {
-        if (auto constant_value = access.constant_value()) {
-          CARBON_ASSIGN_OR_RETURN(
-              Nonnull<const Value*> instantiated,
-              InstantiateType(*constant_value, access.source_loc()));
-          return todo_.FinishAction(instantiated);
+      } else if (act.pos() == 1) {
+        auto constant_value = access.constant_value();
+        if (constant_value) {
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp:431]: nit: Prefer the "if with initializer" form to avoid bringing that variable into scope. Same applies to similar snippets in the rest of the code

File: explorer/interpreter/interpreter.cpp [hunk 5]
```diff
@@ -423,30 +424,32 @@ auto Interpreter::StepLvalue() -> ErrorOr<Success> {
...
+          return todo_.Spawn(std::make_unique<TypeInstantiationAction>(
+              *constant_value, access.source_loc()));
         }
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp:439]: I think 
```
} else {
  ...
  return ...
}
```
might be easier to parse here, otherwise it looks like we do 2 `FinishAction` one after the other at a glance (even if that's not the case).

Same applies to similar code in your PR

File: explorer/interpreter/interpreter.cpp [hunk 6]
```diff
@@ -457,19 +460,20 @@ auto Interpreter::StepLvalue() -> ErrorOr<Success> {
...
-      } else {
-        // Append `.base` element to the address, and return the new LValue.
-        Address object = cast<LValue>(*act.results()[0]).address();
-        Address base = object.ElementAddress(&access.element());
-        return todo_.FinishAction(arena_->New<LValue>(base));
       }
+      // Append `.base` element to the address, and return the new LValue.
+      Address object = cast<LValue>(*act.results()[0]).address();
+      Address base = object.ElementAddress(&access.element());
+      return todo_.FinishAction(arena_->New<LValue>(base));
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: Might want to revert that one, except if there was a specific reason behind it?

File: explorer/interpreter/interpreter.cpp (hunk 3 above)
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp]: I think that for `EvalAssociatedConstant` specifically we can preserve the existing flow (without spawning an action), given this is part of the type instantiation. That's a perfectly fine first step / PR at the very least to me.

File: explorer/interpreter/interpreter.cpp [hunk 7]
```diff
@@ -1278,36 +1309,52 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
+                 act.pos() == 2) {
         // Next, if we're accessing an interface member, evaluate the `impl`
         // expression to find the corresponding witness.
-        return todo_.Spawn(
-            std::make_unique<WitnessAction>(access.impl().value()));
+        if (act.pos() == 1 && access.impl().has_value() &&
+            !forming_member_name) {
+          return todo_.Spawn(
+              std::make_unique<WitnessAction>(access.impl().value()));
+        }
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: Recommended here:
```
} else if (...) {
  ...
} else if (...) { ...
```

File: explorer/interpreter/interpreter.cpp [hunk 8]
```diff
@@ -1278,36 +1309,52 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
+        // First evaluate the first operand
         if (access.is_addr_me_method()) {
           return todo_.Spawn(std::make_unique<LValAction>(&access.object()));
         } else {
           return todo_.Spawn(
               std::make_unique<ExpressionAction>(&access.object()));
         }
-      } else if (act.pos() == 1 && access.impl().has_value() &&
-                 !forming_member_name) {
+      } else if ((act.pos() == 1 && access.impl().has_value() &&
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: I'm not sure that flow makes sense anymore if we end up diverging for all these subcases. The logic becomes difficult to understand. Would that look better if we invert the "step" and "target" conditions?

For example by doing (pseudocode)
```
  else {
    if (access.impl().has_value() && !forming_member_name) {
       if (act.pos() == 1) {
       ...
       } else if (act.pos() == 2) {
       ...
       } else {
          return ...;
       }
    else if (found_in_interface=...; found_in_interface) {
       // Pos 1-n
    } else if () {
      ...
    }
    }
  }
```

Feel free to push back if you believe that makes it even more complicated.

File: explorer/interpreter/interpreter.cpp (hunk 3 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: It is used in the same way in `InstantiateBindings`, `ConvertStructToClass` and `CallFunction` as well.
So, should we preserve the existing flow for these three methods as well?

File: explorer/interpreter/interpreter.cpp (hunk 6 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: oh yes I will revert it.

File: explorer/interpreter/interpreter.cpp (hunk 4 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp:431]: Oh sure I'll do that

File: explorer/interpreter/interpreter.cpp (hunk 5 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp:439]: Can I write it like this? but here I didn't use if with initializer for `constant_value` as it is used at two places.
```cpp
    case ExpressionKind::SimpleMemberAccessExpression: {
      const auto& access = cast<SimpleMemberAccessExpression>(exp);
      const auto constant_value = access.constant_value();
      if (auto rewrite = access.rewritten_form()) {
        return todo_.ReplaceWith(std::make_unique<LValAction>(*rewrite));
      }
      if (act.pos() == 0) {
        //    { {e.f :: C, E, F} :: S, H}
        // -> { e :: [].f :: C, E, F} :: S, H}
        return todo_.Spawn(std::make_unique<LValAction>(&access.object()));
      } else if (act.pos() == 1 && constant_value) {
          return todo_.Spawn(std::make_unique<TypeInstantiationAction>(
              *constant_value, access.source_loc()));
      } else {
        if(constant_value) {
          return todo_.FinishAction(act.results().back());
        } else {
          //    { v :: [].f :: C, E, F} :: S, H}
          // -> { { &v.f :: C, E, F} :: S, H }
          Address object = cast<LValue>(*act.results()[0]).address();
          Address member = object.ElementAddress(&access.member());
          return todo_.FinishAction(arena_->New<LValue>(member));
        }
      }
    }
    ```

File: explorer/interpreter/interpreter.cpp (hunk 8 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: Yes I tried it and certainly makes the code more readable (maybe just for me as I implemented it) but an extra level of nesting it comes with.

File: explorer/interpreter/interpreter.cpp (hunk 3 above)
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp]: If you haven't changed those yet, I would recommend preserving the flow to see what this looks like yes.

File: explorer/interpreter/interpreter.cpp (hunk 5 above)
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp:439]: It tends to be easier to understand when the conditions reflect `act.pos()`, so at a high level you can see 
* Pos1
* Pos2
* ...

If there were more code reuse maybe it would make more sense, but here I think I would still recommend branching by `act.pos()` only.

Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp:439]: When branching with just `act.pos()`, I'm facing a little issue here.
Look that `What should I return here?` comment in the code below. If I don't return anything there will will fall through in the switch statement. How should I handle this case?
```cpp
if (act.pos() == 0) {
        ...
} else if (act.pos() == 1) {
    if(constant_value) {
        return todo_.Spawn(std::make_unique<TypeInstantiationAction>(
           *constant_value, access.source_loc()));
    } else {
       // What should I return here?
    }
} else {
       ...
}
```

Comment (Pixep) [reply explorer/interpreter/interpreter.cpp:439]: That would be `RunAgain()` precisely!

Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp:439]: Oh, I got it!! Thanks.

File: explorer/interpreter/interpreter.cpp [hunk 9]
```diff
@@ -414,37 +416,42 @@ auto Interpreter::StepLvalue() -> ErrorOr<Success> {
...
+          return todo_.FinishAction(arena_->New<LValue>(member));
         }
-        //    { v :: [].f :: C, E, F} :: S, H}
-        // -> { { &v.f :: C, E, F} :: S, H }
-        Address object = cast<LValue>(*act.results()[0]).address();
-        Address member = object.ElementAddress(&access.member());
-        return todo_.FinishAction(arena_->New<LValue>(member));
       }
     }
+
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: Probably unintended change?
```suggestion
```

File: explorer/interpreter/interpreter.cpp [hunk 10]
```diff
@@ -462,7 +469,8 @@ auto Interpreter::StepLvalue() -> ErrorOr<Success> {
       if (act.pos() == 0) {
         // Get LValue for expression.
         return todo_.Spawn(std::make_unique<LValAction>(&access.object()));
-      } else {
+      }
+      {
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: Likely unintended as well
```suggestion
      } else {
```

File: explorer/interpreter/interpreter.cpp [hunk 11]
```diff
@@ -1075,6 +1083,46 @@ auto Interpreter::CallFunction(const CallExpression& call,
   }
 }
 
+// -
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
```

File: explorer/interpreter/interpreter.cpp [hunk 12]
```diff
@@ -1075,6 +1083,46 @@ auto Interpreter::CallFunction(const CallExpression& call,
...
+auto Interpreter::StepInstantiateType() -> ErrorOr<Success> {
+  Action& act = todo_.CurrentAction();
+  Nonnull<const Value*> type = cast<TypeInstantiationAction>(act).type();
+  SourceLocation source_loc = cast<TypeInstantiationAction>(act).source_loc();
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp:1066]: If the two above are not mutated, could be good to qualify them as `const`

File: explorer/interpreter/interpreter.cpp [hunk 13]
```diff
@@ -1283,15 +1370,14 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
             std::make_unique<ExpressionAction>(&access.object()));
-      } else {
-        ElementPath::Component base_elt(&access.element(), std::nullopt,
-                                        std::nullopt);
-        const Value* value = act.results()[0];
-        CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> base_value,
-                                value->GetElement(arena_, ElementPath(base_elt),
-                                                  exp.source_loc(), value));
-        return todo_.FinishAction(base_value);
       }
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: Probably unwanted `} else {` removal

File: explorer/interpreter/interpreter.cpp [hunk 14]
```diff
@@ -1075,6 +1083,46 @@ auto Interpreter::CallFunction(const CallExpression& call,
...
 }
 
+// -
+auto Interpreter::StepInstantiateType() -> ErrorOr<Success> {
+  Action& act = todo_.CurrentAction();
+  Nonnull<const Value*> type = cast<TypeInstantiationAction>(act).type();
+  SourceLocation source_loc = cast<TypeInstantiationAction>(act).source_loc();
+
+  switch (type->kind()) {
+    case Value::Kind::NominalClassType: {
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp:1069]: Is it worth redefining those, in addition to the ones in `InstantiateType`? I'm slightly worried we might be duplicating code between both

File: explorer/interpreter/action.h [hunk 15]
```diff
@@ -198,6 +199,26 @@ class ExpressionAction : public Action {
   Nonnull<const Expression*> expression_;
 };
 
+class TypeInstantiationAction : public Action {
```
Comment (Pixep) [inline explorer/interpreter/action.h:204]: Can you add a little comment above?

File: explorer/interpreter/interpreter.cpp [hunk 16]
```diff
@@ -1141,68 +1188,82 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
+            std::optional<Nonnull<const Witness*>> witness;
+            if (access.impl().has_value()) {
+              witness = cast<Witness>(act.results()[1]);
+            }
+            ElementPath::Component member(&access.member(), found_in_interface,
+                                          witness);
+            const Value* aggregate;
+            if (access.is_type_access()) {
+              CARBON_ASSIGN_OR_RETURN(
+                  aggregate, InstantiateType(&access.object().static_type(),
```
Comment (Pixep) [inline explorer/interpreter/interpreter.cpp]: Ha, I suppose that's another opportunity to use the Action you created. Hopefully that makes the rest a bit more readable / symmetric, GH makes it hard to parse the diff.

**Review by Pixep — CHANGES_REQUESTED** [review]
Some general comments, and I think we are far enough to invite @zygoloid to have a look. The diff is fairly large at places, but shouldn't be surprising given the change I think.

File: explorer/interpreter/interpreter.cpp [hunk 17]
```diff
@@ -1141,68 +1188,82 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
+          auto found_in_interface = access.found_in_interface();
+          if (act.pos() == 1 && found_in_interface) {
+            return todo_.Spawn(std::make_unique<TypeInstantiationAction>(
+                *found_in_interface, exp.source_loc()));
+          } else {
+            if (found_in_interface) {
+              found_in_interface = cast<InterfaceType>(act.results().back());
+            }
+            CARBON_CHECK(phase() == Phase::CompileTime)
+                << "should not form MemberNames at runtime";
```
Comment (zygoloid) [inline explorer/interpreter/interpreter.cpp]: It'd be clearer to move this check up to the start of this `if (... = dyn_cast<TypeOfMemberName>(...))` block.

File: explorer/interpreter/interpreter.cpp [hunk 18]
```diff
@@ -1217,64 +1278,90 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
-            CARBON_CHECK(access.member().base_type().has_value())
-                << "compound access should have base type or impl";
+            // If we're forming a member name, we must be in the outer
+            // evaluation in `Type.(Interface.method)`. Produce the same method
+            // name with its `type` field set.
+            if (found_in_interface) {
+              found_in_interface = cast<InterfaceType>(act.results().back());
+            }
+            CARBON_CHECK(phase() == Phase::CompileTime)
+                << "should not form MemberNames at runtime";
```
Comment (zygoloid) [inline explorer/interpreter/interpreter.cpp]: Please move this check up to the top of this `if (forming_member_name)` block.

**Review by zygoloid — COMMENTED** [review]
This is looking great!

It looks like after this change we have two calls to `InstantiateType` left that still need to get converted over to the new action. I think it's fine to land this before doing that, and finish it off and remove `InstantiateType` as a separate change if you prefer -- or to do it all as part of this PR.

File: explorer/interpreter/interpreter.cpp (hunk 14 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp:1069]: Oh I didn't notice that, I think we should remove those two from `InstantiateType` method as both `NominalClassType` and `PointerType` need to instantiate the type again and it should spawn an action but in `InstantiateType` method it directly evaluates by calling `InstantiateType` directly.

What do you suggest?

File: explorer/interpreter/interpreter.cpp (hunk 17 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: Sure, I'll do that.

File: explorer/interpreter/interpreter.cpp (hunk 16 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: Sorry, I missed it by mistake, making it similar to how I did it for case `CompoundMemberAccessExpression`

File: explorer/interpreter/interpreter.cpp (hunk 18 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: Sure, I'll do that

File: explorer/interpreter/interpreter.cpp (hunk 13 above)
Comment (prabhatexit0) [reply explorer/interpreter/interpreter.cpp]: Sorry I did this by mistake, reverting it in next commit.

**Comment by prabhatexit0** [comment]
> This is looking great!
> 
> It looks like after this change we have two calls to `InstantiateType` left that still need to get converted over to the new action. I think it's fine to land this before doing that, and finish it off and remove `InstantiateType` as a separate change if you prefer -- or to do it all as part of this PR.

Thanks Zygoloid,
regarding `InstantiateType`. I'm not able to figure out how I should use the **new flow** (Instantiating type using action) in methods like `Interpreter::EvalAssociatedConstant`, `Interpreter::InstantiateBindings` as these are not transition methods for Actions. I'll need a bit of guidance on how I should approach it.
I am fine with doing this change as part of this PR but I would prefer to do it separately, this is my first time doing an open source contribution and think I did some mistakes like not creating a separate branch on forked repository, a bit of mess in commits etc.
apologies for that, but I'll learn from it and hopefully not repeat the same mistakes again.

File: explorer/interpreter/interpreter.cpp (hunk 14 above)
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp:1069]: That's probably the right approach is we lean towards a `Step` based intantiation, yes!

**Review by aryuemaan — APPROVED** [review]

**Comment by Pixep** [comment]
> > This is looking great!
> > It looks like after this change we have two calls to `InstantiateType` left that still need to get converted over to the new action. I think it's fine to land this before doing that, and finish it off and remove `InstantiateType` as a separate change if you prefer -- or to do it all as part of this PR.
> 
> Thanks Zygoloid, regarding `InstantiateType`. I'm not able to figure out how I should use the **new flow** (Instantiating type using action) in methods like `Interpreter::EvalAssociatedConstant`, `Interpreter::InstantiateBindings` as these are not transition methods for Actions. I'll need a bit of guidance on how I should approach it. I am fine with doing this change as part of this PR but I would prefer to do it separately, this is my first time doing an open source contribution and think I did some mistakes like not creating a separate branch on forked repository, a bit of mess in commits etc. apologies for that, but I'll learn from it and hopefully not repeat the same mistakes again.

No problem with keeping this level of change in for this pull request, that is also what @zygoloid suggested.
I think that's great. We will use the `description` of this PR in the commit message: could you update it to reflect what you have done here? Some recommendations:

* removing the image, and initial context/questions that are not anymore relevant (resolved, done differently, ...), 
* changing "Close #2594" to `Relates to #2594" (as we're not completely done yet),
* and overall just having a simple explanation of what your contributions does and how (feel free to keep it simple).

Thank you!

**Comment by prabhatexit0** [comment]
> > > This is looking great!
> > > It looks like after this change we have two calls to `InstantiateType` left that still need to get converted over to the new action. I think it's fine to land this before doing that, and finish it off and remove `InstantiateType` as a separate change if you prefer -- or to do it all as part of this PR.
> > 
> > 
> > Thanks Zygoloid, regarding `InstantiateType`. I'm not able to figure out how I should use the **new flow** (Instantiating type using action) in methods like `Interpreter::EvalAssociatedConstant`, `Interpreter::InstantiateBindings` as these are not transition methods for Actions. I'll need a bit of guidance on how I should approach it. I am fine with doing this change as part of this PR but I would prefer to do it separately, this is my first time doing an open source contribution and think I did some mistakes like not creating a separate branch on forked repository, a bit of mess in commits etc. apologies for that, but I'll learn from it and hopefully not repeat the same mistakes again.
> 
> No problem with keeping this level of change in for this pull request, that is also what @zygoloid suggested. I think that's great. We will use the `description` of this PR in the commit message: could you update it to reflect what you have done here? Some recommendations:
> 
>     * removing the image, and initial context/questions that are not anymore relevant (resolved, done differently, ...),
> 
>     * changing "Close [Explorer: change `InstantiateType` to be an `Action` instead of a method #2594](https://github.com/carbon-language/carbon-lang/issues/2594)" to `Relates to [Explorer: change `InstantiateType` to be an `Action` instead of a method #2594](https://github.com/carbon-language/carbon-lang/issues/2594)" (as we're not completely done yet),
> 
>     * and overall just having a simple explanation of what your contributions does and how (feel free to keep it simple).
> 
> 
> Thank you!

Thanks Pixep,
I changed the PR description.

**Review by Pixep — APPROVED** [review]
Great, thanks a lot @prabhatexit0 ! I'm happy with that :).
@zygoloid would you mind having a last look when you have time?

**Comment by prabhatexit0** [comment]
> Great, thanks a lot @prabhatexit0 ! I'm happy with that :). @zygoloid would you mind having a last look when you have time?

@Pixep Welcome and Thank you so much!

**Comment by zygoloid** [comment]
> @zygoloid would you mind having a last look when you have time?

Looks great to me.
@prabhatexit0 Thank you for your work on this!
